src/loc.rs

```diff
@@ -0,0 +1,466 @@
+        if location.range.is_end() {
+            self.input = EndianBuf::new(&[]);
+        } else {
+            self.input = rest;
+        }
+
+        Ok(Some(location))
+    }
+}
+
+impl<'input, Endian> FallibleIterator for RawLocationListIter<'input, Endian>
+    where Endian: Endianity
+{
+    type Item = LocationListEntry<'input, Endian>;
+    type Error = Error;
+
+    fn next(&mut self) -> ::std::result::Result<Option<Self::Item>, Self::Error> {
+        RawLocationListIter::next(self)
+    }
+}
+
+/// An iterator over a location list.
+#[derive(Debug)]
+pub struct LocationListIter<'input, Endian>
+    where Endian: Endianity
+{
+    raw: RawLocationListIter<'input, Endian>,
+    base_address: u64,
+}
+
+impl<'input, Endian> LocationListIter<'input, Endian>
+    where Endian: Endianity
+{
+    /// Construct a `LocationListIter`.
+    pub fn new(input: EndianBuf<'input, Endian>,
+               address_size: u8,
+               base_address: u64)
+               -> LocationListIter<'input, Endian> {
+        LocationListIter {
+            raw: RawLocationListIter::new(input, address_size),
+            base_address: base_address,
+        }
+    }
+
+    /// Advance the iterator to the next location.
+    pub fn next(&mut self) -> Result<Option<LocationListEntry<'input, Endian>>> {
+        loop {
+            let mut location = match try!(self.raw.next()) {
+                Some(location) => location,
+                None => return Ok(None),
+            };
+
+            if location.range.is_end() {
+                return Ok(None);
+            }
+
+            if location.range.is_base_address(self.raw.address_size) {
+                self.base_address = location.range.end;
+                continue;
+            }
+
+            if location.range.begin == location.range.end {
+                // An empty location list entry, skip it.
+                continue;
+            }
+
+            location.range.add_base_address(self.base_address, self.raw.address_size);
+            if location.range.begin > location.range.end {
+                self.raw.input = EndianBuf::new(&[]);
+                return Err(Error::InvalidLocationAddressRange);
+            }
+
+            return Ok(Some(location));
+        }
+    }
+}
+
+impl<'input, Endian> FallibleIterator for LocationListIter<'input, Endian>
+    where Endian: Endianity
+{
+    type Item = LocationListEntry<'input, Endian>;
+    type Error = Error;
+
+    fn next(&mut self) -> ::std::result::Result<Option<Self::Item>, Self::Error> {
+        LocationListIter::next(self)
+    }
+}
+
+/// A location list entry from the `.debug_loc` section.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct LocationListEntry<'input, Endian>
+    where Endian: Endianity
+{
+    /// The address range that this location is valid for.
+    pub range: Range,
+
+    /// The data containing a single location description.
+    pub data: EndianBuf<'input, Endian>,
```

Comment:
I slightly prefer keeping members private and having getter methods. It allows the most flexibility on our end and isn't really any less ergonomic for users. The one thing is that it disallows modifying the members, which is desirable IMO.

Is there a reason you preferred public members here?